Game

Hemmemm opened the pull request.

Hemmemm committed:
Added base implementation

Hemmemm committed:
Deleted useless stuff (issue #1)

Hemmemm committed:
Added a Makefile and worked on the game structure (issue #1)(closes issue #3)

Hemmemm committed:
Previous commit closes #3

Hemmemm committed:
Started implementing movement (issue #1)

Hemmemm committed:
Continue to implement the Snake class (issue #1)

Hemmemm committed:
Changed operator "<<" overloading (issue #1)

Hemmemm committed:
IMPLEMENTED basic snake movement and drwing in ncurses console

Hemmemm committed:
Fixed main game cycle, added checking for forbidden movement change

Hemmemm committed:
Test version of a game

Hemmemm committed:
Moved from sleep funcs to sleep_for (C++11) (issue #1)

Hemmemm committed:
Further improvments of Makefile(issue #3) and reorganisation

Hemmemm committed:
Competed Makefile (closes #3) and some misc changes

Hemmemm committed:
Cleaned Makefile (issue #3)

Hemmemm committed:
Implemented Ball and done some testing w/ it

Hemmemm committed:
Forgot the Ball in the previous commit

Hemmemm committed:
Added intersections and display for the borders

Hemmemm committed:
Added intersections for snake w/ itself

Hemmemm committed:
Some minor changes

Hemmemm committed:
Changed ESC delay time to be 0 sec

Hemmemm committed:
Made snake go through walls and updated collision detection

gigafiga21 commented on Libraries/Snake/snake.h:
```diff
@@ -0,0 +1,29 @@
+#include "../Ball/ball.h"
+
+extern Point dim;
+
```
Add description of the class properties in JSDoc
```
/**
 * Descriprion of the class
 * @prop {propertyType} propertyName - description of the property
 */
```
JSDoc rules: http://usejsdoc.org

gigafiga21 commented on Libraries/Snake/snake.h:
```diff
@@ -0,0 +1,29 @@
+
+class Snake
+{
+private:
```
#### Code style
Tabulation before "private", "public", etc.

gigafiga21 commented on Libraries/Snake/snake.h:
```diff
@@ -0,0 +1,29 @@
+class Snake
+{
+private:
+	short direction;
```
#### Code style
Its better to use speses instead of tabs (4 spaces = 1 tab)

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,170 @@
+#include "snake.h"
+
```
Add documentation for constants via JSDoc
```
/**
 * @const {constType} - description of the constant
 */
```

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,170 @@
+#define NOWCOLL 3
+#define NOCOLL 4
+#define BALL 0
+
```
Add documentation for functions
```
/**
 * Description of the function (if it is constructor - @constructor)
 * @param {parameterType} parameterName - description of parameter
 * @return  {type}                                             - description (optional)
 */
```

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,170 @@
+	if (abs(dir) == 1 or abs(dir) == 2){
+		short* changable;
+		if (abs(dir) == 1)
+			changable = &begin.x;
```
It is better to use code block, even if it is 1 line under if/for/while

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,170 @@
+#define BALL 0
+
+bool Snake::init(Point begin, short dir, int length){
+	if (abs(dir) == 1 or abs(dir) == 2){
```
Replace 1 and 2 with preprocessor constants

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,170 @@
+			value = 1;
+		else
+			value = -1;
+		for (int i = 0; i < length; i++){
```
'{' on a new line, pls

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,170 @@
+
+bool Snake::init(Point begin, short dir, int length){
+	if (abs(dir) == 1 or abs(dir) == 2){
+		short* changable;
```
Add comment or give understandale name
```
/**
 * Description of the variable
 * @type {variableType}
 */
```

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,170 @@
+			changable = &begin.x;
+		else
+			changable = &begin.y;
+		short value;
```
Ununderstandable name

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,170 @@
+}
+
+
+/**
```
Rewrite in JSDoc where direction variable deligated

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,170 @@
+	p.x = snakeBody.front().x;
+	p.y = snakeBody.front().y;
+	switch (direction){
+		case 1:{
```
Code block is not obligatory here

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,170 @@
+		}
+	}
+	moveHead(p, change);
+	Point g;
```
Ununderstandable name, comment or rename

gigafiga21 commented on Libraries/Game/game.h:
```diff
@@ -0,0 +1,31 @@
+#include <thread>
+#include <random>
+
+extern Point dim;
```
It is better not to use global variables.
Ununderstandable name.

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    initscr();
+
+    nodelay(stdscr, TRUE);
+    set_escdelay(0);
```
Remove symbols shoving in console after ESC is pressed

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    Point screenSize;
+    getmaxyx(stdscr, screenSize.y, screenSize.x);
+
+    Point p, g;
```
Description of the variables

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    int dir = 1, len = 5;
+    initSnake(p, dir, len);
+
+    curs_set(0); //"Убиваем" курсор
```
Unnecessary comments, delete

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    Ball ball(dim);
+    ball.generateBall(labyrinth);
+
+    Point* change[2];
```
Ununderstandable variable, add comment pls

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    change[1] = new Point [1];
+    while ( !flag )
+    {
+
```
Remove empty line

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    while ( !flag )
+    {
+
+    	int ch = getch();
```
Replace name (e.g command)

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    		case ERR:
+    			break;
+    		case KEY_UP:
+    			snake.setDirection(-2);
```
Replace -2, 2, -1, 1 with preprocessor constants

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+                break;
+
+    	}
+        if (flag) break;
```
Code block afte if

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    	clear();
+        displayLabyrinth();
+    	for(auto it = snake.snakeBody.begin(); it != snake.snakeBody.end(); it++){
+    		mvaddch((*it).y, (*it).x, '*');
```
Replace (*var).prop with var->prop

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    	flushinp();
+    	mSleep(speed);
+    	refresh();
+        if (flag) break;
```
Duplicate of 63. Is it necessary?

Hemmemm replied:
it is necessary because right after the first use it gets assigned to the snake.move() result
And if the snake 'dies' during the move we still exit the game loop

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    return 0;
+}
+
+void Game::mSleep(int time){
```
Move wrapper into common.cpp, refuse from threads

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,132 @@
+    }
+}
+
+void Game::updateLabyrinth(Point* update[2],int size){
```
Space between comma and type

gigafiga21 commented on Libraries/Snake/snake.h:
```diff
@@ -0,0 +1,29 @@
+	void moveHead(Point p, Point* change[2]);
+public:
+	bool init(Point begin, short direction, int length);
+	std::deque<Point> snakeBody;
```
Just body

gigafiga21 commented on Libraries/Common/common.cpp:
```diff
@@ -0,0 +1,21 @@
+#ifndef COMMON_H
+#define COMMON_H
+#include <iostream>
+#include <deque>
```
Unnecessary include

gigafiga21 commented on Libraries/Ball/ball.cpp:
```diff
@@ -0,0 +1,35 @@
+typedef std::mt19937 MyRNG;
+uint32_t seed_val;
+
+MyRNG rng;
```
Why it is global?

gigafiga21 commented on Libraries/Ball/ball.cpp:
```diff
@@ -0,0 +1,35 @@
+Ball::Ball(Point dim){
+	seed_val = time(NULL);
+    rng.seed(seed_val);
+	std::uniform_int_distribution<int> distribution1(0, dim.x - 1);
```
Move into initializer list

gigafiga21 commented on Libraries/Ball/ball.cpp:
```diff
@@ -0,0 +1,35 @@
+
+bool Ball::generateBall(bool** labyrinth){
+	Point p;
+	//auto dice = std::bind ( distribution, generator );
```
Remove

gigafiga21 commented on Libraries/Ball/ball.cpp:
```diff
@@ -0,0 +1,35 @@
+bool Ball::generateBall(bool** labyrinth){
+	Point p;
+	//auto dice = std::bind ( distribution, generator );
+	//Generate numbers until u get a free spot
```
Multiline comment
```
/**
 *
 */
```

Hemmemm committed:
Switched to using spaces as method of indentation

Hemmemm committed:
Added comments, fixed the issues

Hemmemm committed:
Changed blocks so the opening braket in now on a separate line

Hemmemm committed:
Fixed copying of the snake body, made some explanatory comments for the variables

gigafiga21 commented on Libraries/Ball/ball.cpp:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * @constructor
+ * @param {Point} dim - dimensions of the game field
```
Dimentions is 1, 2, 3, etc. Maybe fieldSize or just size?

gigafiga21 commented on Libraries/Ball/ball.cpp:
```diff
@@ -0,0 +1,50 @@
+{
+    seed_val = time(NULL);
+    rng.seed(seed_val);
+    std::uniform_int_distribution<int> distribution1(0, dim.x - 1);
```
distribution1 and distribution2 aren't necessary:
`distributionX = *(new std::uniform_int_distribution<int>(0, fieldSize.x - 1));`

gigafiga21 commented on Libraries/Ball/ball.cpp:
```diff
@@ -0,0 +1,50 @@
+}
+
+/**
+ * @function Generate the Ball
```
`@function` marks should be on separate line. Wrong mark, right is `@method`. It is not obligatory here.

gigafiga21 commented on Libraries/Ball/ball.cpp:
```diff
@@ -0,0 +1,50 @@
+/**
+ * @function Generate the Ball
+ * @param {bool**} labyrinth -  2-dimensional array defying current state of every point of the game field (blocked or not)
+ * @return {bool} mark of whether the ball was successfully generated
```
```
* @param  {bool**} labyrinth -  2-dimensional array defying current state of every point of the game field (blocked or not)
* @return {bool}             - mark of whether the ball was successfully generated
```

gigafiga21 commented on Libraries/Ball/ball.cpp:
```diff
@@ -0,0 +1,50 @@
+    /**
+     * Generate numbers until u get a free spot
+     */
+    while (1){
```
`\n{`

gigafiga21 commented on Libraries/Common/common.cpp:
```diff
@@ -0,0 +1,48 @@
+ */
+inline void mSleep(int time)
+{
+    std::this_thread::sleep_for(std::chrono::milliseconds(time));
```
`usleep(milliseconds)` for linux, `sleep(milliseconds)` for windows.

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+#include "game.h"
+
+/**
+ * @global {Point} gameFieldSize - a variable, which stores the dimensions of the game field
```
```
/**
 * Description
 * @type {varType}
 * @global
 */
```

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+    gameFieldSize.x = size;
+    gameFieldSize.y = size/2;
+    setLabyrinth(gameFieldSize);
+    setSpeed(sp);
```
Rename sp into speed

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+
+/**
+ * @function Main game loop
+ * @return {int} the exit code
```
'-' between {type} and description

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+    change[1] = new Point [1];
+
+    std::list<Point> currBody;
+    while ( !flag )
```
Remove spaces around `!flag`

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+        clear();
+        displayLabyrinth();
+        snake.getCoords(&currBody);
+        for(auto it = currBody.begin(); it != currBody.end(); it++)
```
It's better not to redraw all snake (at Windows console this algorithm will produce blinking), just erase last cell and draw new first

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+        flag = snake.move(labyrinth, &ball, change);
+        updateLabyrinth(change, 1);
+        clear();
+        displayLabyrinth();
```
Why are you redrawing labirynth here? It is enough to redraw last and first cells of snake and change 2 cells in the array of labyrinth.

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+
+/**
+ * @function Set's the starting labyrinth
+ * @param {Point} dimensions - dimensions of the game field
```
Aaargh, this is size

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+ *                 update[1] - an array containing Points to remove from the labyrinth
+ * @param {int} size - the longest sequence for updating [max(len(update[0], update[1]))]
+ */
+void Game::updateLabyrinth(Point* update[2],int size)
```
Space between ',' and type name

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+ * @param {int} sp - speed of the game(1 / refresh rate) in milliseconds
+ */
+void Game::setSpeed(int sp)
+
```
Remove this empty line

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+ * @function The method to change the game speed without giving the direct access
+ * @param {int} sp - speed of the game(1 / refresh rate) in milliseconds
+ */
+void Game::setSpeed(int sp)
```
Rename sp into speed

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,203 @@
+void Game::setSpeed(int sp)
+
+{
+    speed = sp;
```
`this->speed = speed;`

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,245 @@
+            snakeBody.push_front(begin);
+            *changable += value;
+        }
+        direction = dir;
```
`this->direction = direction`

gigafiga21 commented on Libraries/Snake/snake.h:
```diff
@@ -0,0 +1,64 @@
+#define MVUP -2
+
+/**
+ * @global {Point} gameFieldSize - a variable, which stores the dimensions of the game field
```
Size of the game field

Hemmemm committed:
Reworked displaying the game, some misc fixes and changes

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,217 @@
+ */
+int Game::run()
+{
+
```
Remove this empty line

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,217 @@
+    displayLabyrinth();
+    snake.getCoords(&currBody);
+    for(auto it = currBody.begin(); it != currBody.end(); it++)
+        {
```
Remove extra padding

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,217 @@
+            break;
+        }
+        initChange(change, changeSize);
+        flag = snake.move(labyrinth, &ball, change) == 1;
```
For what purpose? Flag variable not used after.

Hemmemm replied:
it is used at the top, when the !flag statement is checked
but as soon as it is called after the iteration it still lets the game to display correctly

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,217 @@
+    }
+    return q;
+}
+
```
Add documentation

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,217 @@
+    bool q = snake.init(begin, dir, length);
+    std::list<Point> currBody;
+    snake.getCoords(&currBody);
+    if (q){
```
`\n{`

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,247 @@
+ * @param {Point*} change    - 2-dimensional array of changes needed to be applied to the labyrinth
+ *                 change[0] - an array containing Points to add to the labyrinth
+ *                 change[1] - an array containing Points to remove from the labyrinth
+ * @return {short}           - mark of whether the was a non-boundary non-ball collision
```
the was

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,247 @@
+ * The method to set the direction where the snake is heading
+ * @param {int} newDirection - direction of the snake we are trying to set
+ */
+void Snake::setDirection(int newDirection)
```
int direction

Hemmemm committed:
Changed comments, added new, wrapped some parts of the code into functions

Hemmemm committed:
Fixed the comments(hopefully), bugfix

Fixed the bug where the Ball could be spawned int the Position where snake's head just moved due to checking labyrinth before it's update

Hemmemm committed:
Split common in 2 files(.cpp & .h)

Mainly to avoid using inline functions & for cleaner code

Hemmemm committed:
Added common.h

gigafiga21 commented on Libraries/Common/common.cpp:
```diff
@@ -0,0 +1,81 @@
+/**
+ * Operation '<<' override for the Point type
+ * @param  {std::ostream&} s - current ostream variable
+ * @param  {Point}         p - 
```
Add description for p

gigafiga21 commented on Libraries/Common/common.cpp:
```diff
@@ -0,0 +1,81 @@
+}
+
+/**
+ * Operation '<<' override for the Point type
```
Refresh comment

gigafiga21 commented on Libraries/Common/common.cpp:
```diff
@@ -0,0 +1,81 @@
+
+/**
+ * Check if the Point is in the removal queue
+ *
```
Remove

gigafiga21 commented on Libraries/Snake/snake.cpp:
```diff
@@ -0,0 +1,258 @@
+ * Check given by parameters point on Wall collision
+ * @param  {Point}  coords    - coordinates(x, y) of the segment
+ * @param  {bool**} labyrinth -  2-dimensional array defying current state of every point of the game field (blocked or not)
+ * @return {short}           - a type of an intersection
```
(:~{)

Hemmemm committed:
Fixed the comments and removed blank lines from them

Hemmemm committed:
Fixed the comments

gigafiga21 commented on Libraries/Common/common.cpp:
```diff
@@ -0,0 +1,83 @@
+
+/**
+ * Check if Points is in the addition queue
+ *
```
Remove empty line

gigafiga21 commented on Libraries/Common/common.h:
```diff
@@ -0,0 +1,48 @@
+} PointStyle;
+
+std::ostream& operator << (std::ostream &s, Point p);
+
```
Remove empty lines

gigafiga21 commented on Libraries/Game/game.cpp:
```diff
@@ -0,0 +1,256 @@
+
+/**
+ * Initialization of the change array
+ *
```
Remove empty line

Hemmemm committed:
Merge branch 'game' of https://github.com/echo-team/Snoke into game

Hemmemm committed:
Fixed spacing between prototypes

Hemmemm committed:
Added zeroing for the labyrinth

Hemmemm committed:
Moved initialization of ncurses to the Main.cpp

Hemmemm committed:
Moved console-setting functions back to Game::run

gigafiga21 committed:
Merge branch 'master' into game
